README.md
FilmAffinity::UserRating
========================

[![Build Status](https://travis-ci.org/williambelle/filmaffinity-userrating.svg?branch=master)](https://travis-ci.org/williambelle/filmaffinity-userrating)
[![Coverage Status](https://coveralls.io/repos/williambelle/filmaffinity-userrating/badge.svg?branch=master&service=github)](https://coveralls.io/github/williambelle/filmaffinity-userrating?branch=master)
[![CPAN Version](https://img.shields.io/cpan/v/FilmAffinity-UserRating.svg)](https://metacpan.org/release/FilmAffinity-UserRating)

Perl interface to FilmAffinity

Install
-------

Via CPAN with:

```bash
cpan install FilmAffinity::UserRating
```

Usage
-----

### Command Line

##### `filmaffinity-get-ratings.pl`

```console
filmaffinity-get-ratings.pl
Usage:
    filmaffinity-get-rating.pl --userid=123456

    filmaffinity-get-rating.pl --userid=123456 --delay=2

    filmaffinity-get-rating.pl --userid=123456 --output=/home/william/myvote.list

Options:
  --delay=3
    delay between requests

  --output=/home/william/rating.list
    output file
```

##### `filmaffinity-get-movie-info.pl`

```console
filmaffinity-get-movie-info.pl
Usage:
    filmaffinity-get-movie-info.pl --id=123456

    filmaffinity-get-movie-info.pl --id=123456 --delay=2

    filmaffinity-get-movie-info.pl --id=932476 --output=/home/william/matrix.json

Options:
  --delay=3
    delay between requests

  --output=/home/william/matrix.json
    output json file
```

##### `filmaffinity-get-all-info.pl`

```console
filmaffinity-get-all-info.pl
Usage:
    filmaffinity-get-all-info.pl --userid=123456 --destination=path/to/my/folder

    filmaffinity-get-all-info.pl --userid=123456 --destination=path/to/my/folder --delay=2

    filmaffinity-get-all-info.pl --userid=123456 --destination=path/to/my/folder --force

Options:
  --delay=3
    delay between requests

  --force
    force to retrieve all movies
```

### Module

```perl
use FilmAffinity::UserRating;

my $parser = FilmAffinity::UserRating->new( userID => '123456' );
my $ref_movies = $parser->parse();
```

Retrieve information about a filmaffinity movie

```perl
use FilmAffinity::Movie;

my $movie = FilmAffinity::Movie->new( id => '348488' );
$movie->parse();

my $title = $movie->title;
```

Contributing
------------

Contributions are always welcome.

See [Contributing](CONTRIBUTING.md).

Developer
---------

  * [William Belle](https://github.com/williambelle)

License
-------

Copyright (C) 2013-2018 William Belle

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


Changes
Revision history for FilmAffinity-UserRating

1.00    11 Nov 2018 / 21h00
        Fix delay initialization
        Fix usage example
        Improve README.md
        Add CONTRIBUTING.md
        Update dependencies

0.10    31 Nov 2016 / 11h20
        Improve pod documentation
        Clean code

0.09    22 Nov 2014 / 15h45
        Rewrite XPath expression because of the new layout of filmaffinity
        Improve unit test
        add myrating accessor
        bug-fix: unit test failures on Windows

0.08    13 Nov 2014 / 20h10
        Update dependencies

0.07    12 Oct 2014 / 16h00
        Rewrite some code because of the new layout of filmaffinity
        Clean lib/*
        bug-fix: films with empty rating
        bug-fix: films with empty vote

0.06    08 Mar 2014 / 20h00
        bug-fix: encoding problems with unit test
        Improve unit test with json files

0.05    28 Sep 2013 / 18h00
        Improve script to retrieve only the difference since last run
        add "force" option to retrieve all movies again

0.04    29 Aug 2013 / 20h30
        bug-fix: films with empty duration
        bug-fix: films with empty topics
        bug-fix: confusion between title and original title
        New script to get information from filmaffinity

0.03    27 Apr 2013 / 16h00
        Retrieve votes
        Retrieve global rating

0.02    20 Apr 2013 / 17h40
        Get information from filmaffinity about a film

0.01    23 Feb 2013 / 18h00
        First release to "master"
        Get filmaffinity voted movies from a user


t/MockSite.pm
package MockSite;

use File::Temp qw/tempdir/;
use File::Copy::Recursive qw/dircopy/;

# copy a local test directory
sub mockLocalSite {
  my $localdir = shift;
  
  my $tmpDir = tempdir(CLEANUP => 1);
  dircopy($localdir, $tmpDir);
  "file://$tmpDir";
}

1;

Build.PL
use 5.006;
use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
  module_name       => 'FilmAffinity::UserRating',
  license           => 'perl',
  dist_author       => q{William Belle <william.belle@gmail.com>},
  dist_version_from => 'lib/FilmAffinity/UserRating.pm',
  build_requires    => {
    'Test::More'       => 0,
    'Test::JSON'       => 0.11,
    'File::Find::Rule' => 0.34,
    'Test::MockObject' => 1.20,
    'Test::LongString' => 0.17,
  },
  requires => {
    'perl'                     => 5.006,
    'Text::Trim'               => 1.02,
    'LWP::RobotUA'             => 6.36,
    'HTML::TreeBuilder'        => 5.07,
    'HTML::TreeBuilder::XPath' => 0.14,
    'Moose'                    => 2.20,
    'MooseX::Privacy'          => 0.05,
    'IO::All'                  => 0.87,
    'Encode'                   => 2.98,
    'JSON'                     => 2.97,
    'List::Compare'            => 0.53,
    'Readonly'                 => 2.05,
    'File::Copy::Recursive'    => 0.44,
    'IO::Interactive'          => 1.02,
  },
  meta_merge => {
    resources => {
      repository =>
        'https://github.com/williambelle/filmaffinity-userrating',
    },
  },
  add_to_cleanup     => ['FilmAffinity-UserRating-*'],
  create_makefile_pl => 'traditional',
);

$builder->create_build_script();


lib/FilmAffinity/Utils.pm
package FilmAffinity::Utils;

use strict;
use warnings;

use Readonly;
use LWP::RobotUA;

=head1 NAME

FilmAffinity::Utils - Utils for FilmAffinity

=head1 VERSION

Version 1.00

=cut

our $VERSION = 1.00;

=head1 SYNOPSIS

  use FilmAffinity::Utils;

  my $string = demoronize($string);
  my $ua = buildRobot($delay);
  my $tsv = data2tsv($data);

=head1 DESCRIPTION

Several utilities functions

=cut

use base 'Exporter';
our @EXPORT_OK = qw/demoronize buildRobot data2tsv/;

Readonly::Scalar my $SECONDS => 60;

=head1 SUBROUTINES/METHODS

=head2 demoronize

Remove funky Windows 'smart' characters.

Most of this code is borrowed from Catalyst::Plugin::Params::Demoronize

=cut

sub demoronize {
  my $str = shift;

  return $str unless defined $str;

  my $demoronizeReplaceMap = {
    q{\x{201A}|\x82}   => q{,},   # 82, SINGLE LOW-9 QUOTATION MARK
    q{\x{201E}|\x84}   => q{,,},  # 84, DOUBLE LOW-9 QUOTATION MARK
    q{\x{2026}|\x85}   => q{...}, # 85, HORIZONTAL ELLIPSIS
    q{\x{02C6}|\x88}   => q{^},   # 88, MODIFIER LETTER CIRCUMFLEX ACCENT
    q{\x{2018}|\x91}   => q{`},   # 91, LEFT SINGLE QUOTATION MARK
    q{\x{2019}|\x92}   => q{'},   # 92, RIGHT SINGLE QUOTATION MARK
    q{\x{201C}|\x93}   => q{"},   # 93, LEFT DOUBLE QUOTATION MARK
    q{\x{201D}|\x94}   => q{"},   # 94, RIGHT DOUBLE QUOTATION MARK
    q{\x{2022}|\x95|•} => q{*},   # 95, BULLET
    q{\x{2013}|\x96}   => q{-},   # 96, EN DASH
    q{\x{2014}|\x97}   => q{-},   # 97, EM DASH
    q{\x{2039}|\x8B}   => q{<},   # 8B, SINGLE LEFT-POINTING ANGLE QUOTATION MARK
    q{\x{203A}|\x9B}   => q{>},   # 9B, SINGLE RIGHT-POINTING ANGLE QUOTATION MARK
  };

  foreach my $replace ( keys %{$demoronizeReplaceMap} ) {
    my $replacement = $demoronizeReplaceMap->{$replace};
    $str =~ s/$replace/$replacement/xmsg;
  }

  return $str;
}

=head2 buildRobot

Returns a LWP::RobotUA for web requests

=cut

sub buildRobot {
  my ($delay) = shift;

  my $ua = LWP::RobotUA->new( "FilmAffinity-Bot/$VERSION", 'me@foo.com' );
  $ua->timeout($SECONDS);
  $ua->env_proxy;
  $ua->delay( $delay / $SECONDS );

  return $ua;
}

=head2 data2tsv

Transform FilmAffinity::UserRating in tab separated value

Returns a tsv string

=cut

sub data2tsv {
  my ($data) = shift;

  my $tsvString;
  foreach my $mov ( sort { p_sortByRatings($data) } keys %{$data} ) {
    $tsvString .= $mov . "\t" . $data->{$mov}{'title'} . "\t"
      . $data->{$mov}{'rating'} . "\n";
  }

  return $tsvString;
}

sub p_sortByRatings {
  my ($ref_movies) = @_;
  return $ref_movies->{$b}{'rating'} <=> $ref_movies->{$a}{'rating'};
}

=head1 AUTHOR

William Belle, C<< <william.belle at gmail.com> >>

=head1 BUGS AND LIMITATIONS

Please report any bugs or feature requests to C<bug-filmaffinity-userrating at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=FilmAffinity-UserRating>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc FilmAffinity::Utils

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=FilmAffinity-UserRating>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/FilmAffinity-UserRating>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/FilmAffinity-UserRating>

=item * Search CPAN

L<http://search.cpan.org/dist/FilmAffinity-UserRating/>

=back

=head1 SEE ALSO

L<http://www.filmaffinity.com>

=head1 LICENSE AND COPYRIGHT

Copyright 2013 William Belle.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1;    # End of FilmAffinity::Utils


lib/FilmAffinity/UserRating.pm
package FilmAffinity::UserRating;

use 5.006;
use strict;
use warnings;

use Encode;
use Readonly;
use Text::Trim;
use LWP::RobotUA;
use HTML::TreeBuilder;
use HTML::TreeBuilder::XPath;

use Moose;
use MooseX::Privacy;

use FilmAffinity::Utils qw/buildRobot demoronize/;

=head1 NAME

FilmAffinity::UserRating - Perl interface to FilmAffinity

=head1 VERSION

Version 1.00

=cut

our $VERSION = '1.00';

=head1 SYNOPSIS

Get filmaffinity voted movies from a user

    use FilmAffinity::UserRating;

    my $parser = FilmAffinity::UserRating->new( userID => '123456' );

    my $ref_movies = $parser->parse();

Via the command-line program filmaffinity-get-ratings.pl

=head1 DESCRIPTION

=head2 Overview

FilmAffinity::UserRating is a Perl interface to FilmAffinity. You can use
this module to retrieve your rated movies.

=head2 Constructor

=over 4

=item new()

Object's constructor. You should pass as parameter the userID

    my $parser = FilmAffinity::UserRating->new( userID => '123456' );

=back

=head2 Options

=over 4

=item delay

Set the minimum delay between requests to the server, in seconds.

    my $parser = FilmAffinity::UserRating->new(
      userID => '123456',
      delay  => 20,
    );

By default, the delay is 5 seconds

=back

=head1 ACCESSORS

=head2 $parser->userID

get userID

=cut

has userID => (
  is       => 'ro',
  isa      => 'Int',
  required => 1,
);

=head2 $parser->username

get username

=cut

has username => (
  is     => 'ro',
  isa    => 'Str',
  writer => 'p_username',
);

=head2 $parser->movies

get movies

=cut

has movies => (
  is      => 'rw',
  isa     => 'HashRef[Str]',
  default => sub { {} },
);

has ua => (
  is      => 'rw',
  isa     => 'LWP::RobotUA',
  traits  => [qw/Private/],
);

Readonly my $SECONDS => 5;
my $RATING_URL   = 'http://www.filmaffinity.com/en/userratings.php?orderby=2&';
my $XPATH_ID     = '//div[@class="user-ratings-movie-item"]/div/@data-movie-id';
my $XPATH_TITLE  = '//div[@class="mc-title"]/a';
my $XPATH_RATING = '//div[@class="ur-mr-rat"]';

sub BUILD {
  my ($self, $args) = @_;

  $self->ua( buildRobot( $args->{delay} || $SECONDS ) );

  return;
}

=head1 SUBROUTINES/METHODS

=head2 $parser->parse()

This function parses all rating pages of filmaffinity from a user.

    my $ref_movies = $parser->parse();

=cut

sub parse {
  my $self = shift;

  my ($next, $page) = (1, 1);
  while ( $next ){
    my $url = $self->p_buildUrl( $page );
    my $response = $self->ua->get($url);
    if ($response->is_success){
      my $content = $response->decoded_content();
      $self->parseString($content);
      $next = $self->p_isNextPage($content);
      $page++;
    } else {
      $next = 0;
    }
  }
  return $self->movies;
}

=head2 $parser->parseString($content)

This function parses a page of filmaffinity that is available as
a single string in memory.

    $parser->parseString($content);

=cut

sub parseString {
  my ($self, $content) = @_;

  $content = demoronize($content);
  $content = decode('iso8859-1', $content);

  my $tree = HTML::TreeBuilder->new();
  $tree->parse($content);
  $self->p_username($tree->findvalue( '//span[@id="nick"]' ));

  my @ids       = $tree->findnodes_as_strings( $XPATH_ID );
  my @titles    = $tree->findnodes_as_strings( $XPATH_TITLE );
  my @ratings   = $tree->findnodes_as_strings( $XPATH_RATING );

  $self->p_buildMovieInfo(\@ids, \@titles, \@ratings);

  $tree->delete();
  return;
}

private_method p_buildUrl => sub {
  my ($self, $page) = @_;

  return $RATING_URL.'p='.$page.'&user_id='.$self->userID;
};

private_method p_buildMovieInfo => sub {
  my ($self, $ref_ids, $ref_titles, $ref_ratings) = @_;

  for my $i (0..(@{$ref_titles}-1)){
    $self->movies->{${$ref_ids}[$i]} = {
      'title'  => demoronize(${$ref_titles}[$i]),
      'rating' => ${$ref_ratings}[$i],
    }
  }
};

private_method p_isNextPage => sub {
  my ($self, $content) = @_;

  if ($content =~ m/>&gt;&gt;<\/a><\/div><\/div>/xms){
    return 1;
  }
  return 0;
};

=head1 AUTHOR

William Belle, C<< <william.belle at gmail.com> >>

=head1 BUGS AND LIMITATIONS

Please report any bugs or feature requests to C<bug-filmaffinity-userrating at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=FilmAffinity-UserRating>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc FilmAffinity::UserRating


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=FilmAffinity-UserRating>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/FilmAffinity-UserRating>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/FilmAffinity-UserRating>

=item * Search CPAN

L<http://search.cpan.org/dist/FilmAffinity-UserRating/>

=back

=head1 SEE ALSO

L<http://www.filmaffinity.com>

=head1 LICENSE AND COPYRIGHT

Copyright 2013 William Belle.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1; # End of FilmAffinity::UserRating


lib/FilmAffinity/Movie.pm
package FilmAffinity::Movie;

use strict;
use warnings;

use JSON;
use Encode;
use Readonly;
use Scalar::Util qw(looks_like_number);
use Text::Trim;
use LWP::RobotUA;
use HTML::TreeBuilder;
use HTML::TreeBuilder::XPath;

use Moose;
use MooseX::Privacy;

use FilmAffinity::Utils qw/buildRobot/;

=head1 NAME

FilmAffinity::Movie - Perl interface to FilmAffinity

=head1 VERSION

Version 1.00

=cut

our $VERSION = 1.00;

=head1 SYNOPSIS

Retrieve information about a filmaffinity movie

    use FilmAffinity::Movie;

    my $movie = FilmAffinity::Movie->new(
      id    => $movieID,
      delay => $delay || 5,
    );

    $movie->parse();

Via the command-line program filmaffinity-get-movie-info.pl

=head1 DESCRIPTION

=head2 Overview

FilmAffinity::Movie is a Perl interface to FilmAffinity. You can use
this module to retrieve information about a movie.

=head2 Constructor

=over 4

=item new()

Object's constructor. You should pass as parameter the movieID

    my $movie = FilmAffinity::Movie->new( id => '932476' );

=back

=head2 Options

=over 4

=item delay

Set the minimum delay between requests to the server, in seconds.

    my $parser = FilmAffinity::Movie->new(
      userID => '932476',
      delay  => 20,
    );

By default, the delay is 5 seconds

=back

=head1 ACCESSORS

=head2 $movie->id

get id

=head2 $movie->title

get title

=head2 $movie->originaltitle

get original title

=head2 $movie->year

get year

=head2 $movie->duration

get running time (in minutes)

=head2 $movie->synopsis

get synopsis

=head2 $movie->website

get url of the movie website

=head2 $movie->country

get country

=head2 $movie->cover

get url of the cover

=head2 $movie->rating

get site rating

=head2 $movie->myrating

get personal rating

=head2 $movie->votes

get number of votes

=head2 $movie->genre

get genres list

=head2 $movie->topic

get topics list

=head2 $movie->cast

get cast list

=head2 $movie->director

get directors list

=head2 $movie->composer

get composers list

=head2 $movie->screenwriter

get screenwriters list

=head2 $movie->cinematographer

get cinematographers list

=head2 $movie->studio

get studios list

=head2 $movie->producer

get producers list

=cut

has id       => ( is => 'ro', isa => 'Int', required => 1, );
has title    => ( is => 'rw', isa => 'Str', );
has originaltitle    => ( is => 'rw', isa => 'Str', );
has year     => ( is => 'rw', isa => 'Int', );
has duration => ( is => 'rw', isa => 'Int', );
has synopsis => ( is => 'rw', isa => 'Str', );
has website  => ( is => 'rw', isa => 'Str', );
has country  => ( is => 'rw', isa => 'Str', );
has cover    => ( is => 'rw', isa => 'Str', );
has rating   => ( is => 'rw', isa => 'Num', );
has votes    => ( is => 'rw', isa => 'Num', );
has myrating => ( is => 'rw', isa => 'Num', );

has genre => ( is => 'rw', isa => 'ArrayRef[Str]', );
has topic => ( is => 'rw', isa => 'ArrayRef[Str]', );

has cast     => ( is => 'rw', isa => 'ArrayRef[Str]', );
has director => ( is => 'rw', isa => 'ArrayRef[Str]', );
has composer => ( is => 'rw', isa => 'ArrayRef[Str]', );

has screenwriter    => ( is => 'rw', isa => 'ArrayRef[Str]', );
has cinematographer => ( is => 'rw', isa => 'ArrayRef[Str]', );

has studio   => ( is => 'rw', isa => 'ArrayRef[Str]', );
has producer => ( is => 'rw',  );


has tree => (
  is      => 'rw',
  isa     => 'HTML::TreeBuilder',
  traits  => [qw/Private/],
);

has ua => (
  is      => 'rw',
  isa     => 'LWP::RobotUA',
  traits  => [qw/Private/],
);

my $MOVIE_URL     = 'http://www.filmaffinity.com/en/film';
my $XPATH_TITLE   = '//h1[@id="main-title"]';
my $XPATH_RATING  = '//div[@id="movie-rat-avg"]';
my $XPATH_VOTE    = '//div[@id="movie-count-rat"]/span';
my $XPATH_COUNTRY = '//span[@id="country-img"]/img/@title';
my $XPATH_COVER   = '//div[@id="movie-main-image-container"]/a/img/@src';

my @JSON_FIELD = qw(
  id title year synopsis website duration cast director composer screenwriter
  cinematographer genre topic studio producer country cover rating votes
  originaltitle myrating
);

my $FIELD = [
  {
    accessor => 'originaltitle',
    faTag    => 'Original title',
  },
  {
    accessor => 'year',
    faTag    => 'Year',
  },
  {
    accessor => 'synopsis',
    faTag    => 'Synopsis / Plot',
  },
  {
    accessor => 'website',
    faTag    => 'Official Site',
  },
  {
    accessor   => 'duration',
    faTag      => 'Running Time',
    cleanerSub => \&p_cleanDuration,
  },
  {
    accessor   => 'cast',
    faTag      => 'Cast',
    cleanerSub => \&p_cleanPerson,
  },
  {
    accessor   => 'director',
    faTag      => 'Director',
    cleanerSub => \&p_cleanPerson,
  },
  {
    accessor   => 'composer',
    faTag      => 'Music',
    cleanerSub => \&p_cleanPerson,
  },
  {
    accessor   => 'screenwriter',
    faTag      => 'Screenwriter',
    cleanerSub => \&p_cleanPerson,
  },
  {
    accessor   => 'cinematographer',
    faTag      => 'Cinematography',
    cleanerSub => \&p_cleanPerson,
  },
  {
    accessor   => 'genre',
    faTag      => 'Genre',
    cleanerSub => \&p_cleanGenre,
  },
  {
    accessor   => 'topic',
    faTag      => 'Genre',
    cleanerSub => \&p_cleanGenre,
  },
  {
    accessor   => 'studio',
    faTag      => 'Production Co.',
    cleanerSub => \&p_cleanStudio,
  },
  {
    accessor   => 'producer' ,
    faTag      => 'Production Co.',
    cleanerSub => \&p_cleanStudio,
  },
];

Readonly my $SECONDS => 5;

sub BUILD {
  my ($self, $args) = @_;

  $self->tree( HTML::TreeBuilder->new() );
  $self->ua( buildRobot( $args->{delay} || $SECONDS ) );
  return;
}

=head1 SUBROUTINES/METHODS

=head2 $movie->parse()

This method will get the content of the filmaffinity webpage and retrieve
all information about the movie

=cut

sub parse {
  my $self = shift;

  my $content = $self->getContent();
  $self->parsePage($content);
  return;
}


=head2 $movie->getContent()

This method will get the content of the filmaffinity webpage

=cut

sub getContent {
  my ($self) = @_;

  my $url = $self->p_buildUrlMovie($self->id);

  my $response = $self->ua->get($url);
  if ($response->is_success){
    return $response->content();
  }
}

=head2 $movie->parsePage($content)

This method parses a page of filmaffinity that is available as
a single string in memory.

=cut

sub parsePage {
  my ($self, $content) = @_;

  $content = decode('utf-8', $content);
  $self->tree->parse($content);

  foreach my $data (@{$FIELD}){
    $self->p_findField($data);
  }

  $self->p_findRating();
  $self->p_findVotes();
  $self->country( $self->tree->findvalue( $XPATH_COUNTRY ) );
  $self->cover( $self->tree->findvalue( $XPATH_COVER ) );
  $self->title( trim($self->tree->findvalue( $XPATH_TITLE )) );

  $self->tree->delete();
  return;
}



=head2 $movie->toJSON()

This method will export all movie informations in JSON format

=cut

sub toJSON {
  my $self = shift;

  my %data;
  foreach my $field (@JSON_FIELD){
    $data{$field} = $self->$field() if defined $self->$field();
  };

  return to_json(\%data, {pretty => 1});
}

private_method p_findField => sub {
  my ( $self, $data ) = @_;

  my @nodes = $self->tree->findnodes( '//dl[@class="movie-info"]/dt' );
  foreach my $node (@nodes){
    if ( trim( $node->as_text() ) eq $data->{faTag} ){

      my $searched_node = $node->right();

      my $accessor = $data->{accessor};

      my $value = trim( $searched_node->as_text() );

      next if $value eq q{};

      if (defined $data->{cleanerSub}){
        $value = $data->{cleanerSub}($value, $accessor);
      }

      next if not defined $value;

      $self->$accessor( $value );
      last;
    }
  }
};

private_method p_findRating => sub {
  my $self = shift;

  my $rating = $self->tree->findvalue( $XPATH_RATING );

  return if not defined $rating or $rating eq q{};
  $self->rating( trim($rating) );
};

private_method p_findVotes => sub {
  my $self = shift;

  my $votes = $self->tree->findvalue( $XPATH_VOTE );

  return if not defined $votes or $votes eq q{};
  $votes =~ s/\D//gixms;
  $self->votes( trim($votes) );
};

private_method p_buildUrlMovie => sub {
  my ($self, $id) = @_;

  return $MOVIE_URL.$id.'.html';
};

private_method p_removeTextBetweenParenthesis => sub {
  my $content = shift;

  $content =~ s/[(].*[)]//gxms;
  return $content;
};

private_method p_cleanDuration => sub {
  my $value = shift;
  $value =~ s/min[.]//gixms;
  $value = trim($value);

  if ( looks_like_number($value) ){
    return $value;
  }
  return;
};

private_method p_cleanPerson => sub {
  my $value = shift;

  my @persons = split m/,/xms, $value;
  @persons = map {trim (p_removeTextBetweenParenthesis($_) )} @persons;
  return \@persons;
};

private_method p_cleanGenre => sub {
  my ($value, $field) = @_;

  my $pos = $field eq 'genre' ? 0 : 1;
  my @list = split m/[|]/xms, $value;

  if ( defined $list[$pos]){
    my @genres = trim ( split m/[.]/xms, $list[$pos] );
    return \@genres;
  }
  return;
};

private_method p_cleanStudio => sub {
  my ($value, $field) = @_;

  my $pos = $field eq 'studio' ? 0 : 1;
  my @list = split m/[.]\sProducer: /xms, $value;

  my @studio = ();
  if (not defined $list[$pos]){
    return;
  } else {
    @studio = trim ( split m{ / }xms, $list[$pos] );
    return \@studio;
  }
};

=head1 AUTHOR

William Belle, C<< <william.belle at gmail.com> >>

=head1 BUGS AND LIMITATIONS

Please report any bugs or feature requests to C<bug-filmaffinity-userrating at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=FilmAffinity-UserRating>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc FilmAffinity::Movie

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=FilmAffinity-UserRating>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/FilmAffinity-UserRating>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/FilmAffinity-UserRating>

=item * Search CPAN

L<http://search.cpan.org/dist/FilmAffinity-UserRating/>

=back

=head1 SEE ALSO

L<http://www.filmaffinity.com>

=head1 LICENSE AND COPYRIGHT

Copyright 2013 William Belle.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1; # End of FilmAffinity::Movie
